### ISSUE TITLE
[rush] Suggestion: {pre,post}RushLink event hooks

### ISSUE THREAD

**Issue by EdwardDrapkin** [microsoft/rushstack#1729]
There's an issue with the styled-components type definition file in that it depends on the react-native type definition file.  The solution to this is usually to delete the type definition out of node_modules after install, but because linking is a separate step in Rush, `postRushInstall` runs too early, but `preRushBuild` (potentially) runs too late.  A `postRushLink` event would be run exactly at the right time, and if that's added, so should `preRushLink`.  Other uses for this feature might include patching node_modules or running other proprietary modifications to 3rd party packages.

**Comment by b-my** [comment]
@EdwardDrapkin I encountered the same problem. I needed a solution that worked locally and in the ci environment. It's pretty hacky but I just did this:
`    /**
     * The list of shell commands to run after the Rush installation finishes
     * https://github.com/DefinitelyTyped/DefinitelyTyped/issues/33311
     */
    "postRushInstall": [
      "cat /dev/null > common/temp/node_modules/@types/react-native/index.d.ts"
    ],`

Having a `postRushLink` would make for a nicer solution.

**Comment by octogonz** [comment]
We are actually planning to eliminate `rush link` in the future.  We are moving towards the model where `useWorkspaces=true` enabled in **rush.json**, in which case Rush is not involved with creation of symlinks.

The point in time AFTER `rush install` and BEFORE `rush link` is not particularly useful, because the installation is in an incomplete state.  And the effect of running `rush unlink && rush link` is not interestingly different from running `rush install` twice.

Thus I would suggest to amend the design so that:

## `useWorkspaces=false` behavior

- `postRushInstall` runs after `rush link` rather than after `rush install`

I also noticed that if you run `rush install` twice and no work is performed, we still invoke the `postRushInstall` script, which seems inefficient. So we could also change it so that:

- `postRushInstall` is skipped if `rush link` is skipped.

## `useWorkspaces=true` behavior

And for the new modern mode where `rush link` doesn't do anything, the behavior would be:
- `postRushInstall` runs after `rush install`
- `postRushInstall` is skipped if `rush install` is skipped.

@D4N14L does this make sense to you?  It's probably an easy fix.

**Comment by D4N14L** [comment]
Yes this looks fair to me. Internally, we use the `postRushInstall` script to write telemetry about Rush commands, and currently it does report telemetry after a skipped install. It makes sense that this telem would not be reported after this change, as the command was never actually run. Likewise, anything that modifies an install after install has been completed shouldn't need to be run as long as the install itself was never redone.
